xmppserver/src/main/java/org/jivesoftware/openfire/net/RespondingServerStanzaHandler.java

```diff
@@ -92,12 +104,23 @@ protected boolean isStartOfStream(String xml) {
             // Pull namespaces off of the stream:stream stanza and add them to the additional
             List<Namespace> receivedNamespaces = null;
             try {
-                receivedNamespaces = DocumentHelper.parseText(xml + "</stream:stream>").getRootElement().declaredNamespaces();
+                Element rootElement = DocumentHelper.parseText(xml + "</stream:stream>").getRootElement();
+                receivedNamespaces = rootElement.declaredNamespaces();
                 Set<Namespace> additionalNamespaces = receivedNamespaces
                     .stream()
                     .filter(RespondingServerStanzaHandler::isRelevantNamespace)
                     .collect(Collectors.toSet());
                 connection.setAdditionalNamespaces(additionalNamespaces);
+
+                // Create a new session with a new ID if a new stream has started on an existing connection
+                // following TLS negotiation
+                String newStreamId = rootElement.attribute("id").getValue();
+                if (sessionCreated && newStreamId != null) {
+                    ServerSession.AuthenticationMethod existingAuthMethod = session instanceof LocalOutgoingServerSession
+                        ? ((LocalOutgoingServerSession) session).getAuthenticationMethod()
+                        : null;
+                    transferConnectionToNewSession(newStreamId, existingAuthMethod);
```

Comment:
What happens to the _old_ session? Does that linger somewhere? Is that a potential source of duplication, memory leaks, or even (delayed) errors (after something times out / gets cleaned up and triggers data)?

Reply:
Good spot - connection.reinit needed I think so that no references to the session are persisted